### ISSUE TITLE
Document that instr_get_app_pc() may point to a hook: should encode instead of using the bits there

### ISSUE THREAD

**Issue by yangyixiaof** [DynamoRIO/dynamorio#6220]
**Instruction raw bits are wrong **
On Windows 10, I need to print the raw bits of every instruction to a trace file. 
I have successfully done it, but when I check the correctness, I find that after decoding the raw bits into instr_t, the decoded content is not same as the original instr_t. I just decode the raw bits in analysis funciton of instrumentation phase 3 and do not insert any instrumentation when reproducing this error. 
I used DynamoRIO for 2 years, I am not a novice. I do not encode, just decode from original app_pc without any instrumentation happened,  I am sure there is no problem about the re-relativization problem happened in encode.  
This error happens from official release version DynamoRIO-Windows-9.0.1 to nightly build DynamoRIO-Windows-9.93.19503, the newer versions after DynamoRIO-Windows-9.93.19503 are not tested yet. 

Note that the correctness checking code is in the analysis function in instrumentation phase 3, meaning that the analysis function is the first argument of drmgr_register_bb_instrumentation_event. 

I disassemble the error happen original block into a file, the content is as follows:
TAG  0x00007ffd34a2d500
 +0    L3 @0x000002b0eacd9420  4c 8b d1             mov    %rcx -> %r10
 +3    L3 @0x000002b0eac8ad80  b8 23 00 00 00       mov    $0x00000023 -> %eax
 +8    L3 @0x000002b0eac89928  f6 04 25 08 03 fe 7f test   0x7ffe0308[1byte] $0x01
                               01
 +16   L3 @0x000002b0eac8b7a0  75 03                jnz    $0x00007ffd34a2d515
END 0x00007ffd34a2d500
 This is in ntdll.dll and the offset to the start of ntdll.dll is 0x9d500. 
The error is that, in original app_pc for the second instruction, the binary content is 'e9 c1 2c 16 80' which is different from 'b8 23 00 00 00'. This binary content is actually decoded as a jmp instruction instead of a mov instruction. 
I test this error in many files, and that error will definitely happen in all tested files. 
I have quick fixed this error for myself by checking the mov instruction and re-encoding mov instruction into somewhere for myself. 
Only mov instructions in some special blocks in ntdll.dll have this kind of error. 

**To Reproduce**
Write an exe file which is a simple Hello World with simple file read operations. 
```c
int main()
{
    std::string filename = "C:/HomeSpace/file_with_5bytes.bin";

    std::ifstream t;
    t.open(filename.c_str());      // open input file
    t.seekg(0, std::ios::end);    // go to the end
    int length = t.tellg();           // report location (this is the length)
    t.seekg(0, std::ios::beg);    // go back to the beginning
    char* buffer = new char[length];    // allocate memory for a buffer of appropriate dimension
    printf("read buffer start address:%p\n", buffer);
    t.read(buffer, length);       // read the whole file into the buffer
    t.close();                    // close file handle

    if (*(buffer + 0) < 15) {
        std::cout << "run 0th < 15 branch!\n";
    }
    return 0;
}
```
Write client only with the analysis function (first argument of drmgr_register_bb_instrumentation_event) in instrumentation phase 3, the DR_ASSERT in the following code will fail. 
```c
dr_emit_flags_t
app_instruction_analysis_reproduce_bug(void* drcontext, void* tag, instrlist_t* bb, bool for_trace,
	bool translating, void** user_data)
{
	instr_t* first = instrlist_first(bb);
	instr_t* last = instrlist_last(bb);

	// iterate instr
	instr_t* itr = first;
	while (1) {
		app_pc itr_app_pc = instr_get_app_pc(itr);

		app_pc itr_next_app_pc = decode_next_pc(drcontext, itr_app_pc);

		instr_t inst_data{};
		instr_t* inst = &inst_data;
		instr_init(drcontext, inst);

		byte* next_pc = decode(drcontext, itr_app_pc, inst);
		// this assert will fail for some blocks (no more than 20) in ntdll.dll. 
		DR_ASSERT(instr_same(inst, itr));

		instr_free(drcontext, inst);

		if (itr == last) {
			break;
		}
		itr = instr_get_next(itr);
	}

	return DR_EMIT_DEFAULT;
}
```
In dr_client_main function, register above code into drmgr_register_bb_instrumentation_event, the error will happen. 
```
drmgr_register_bb_instrumentation_event(app_instruction_analysis_reproduce_bug, NULL_Function, NULL);
```

Although I have a simple fix (check the mov instruction in ntdll.dll and re-encode mov instruction into somewhere else by myself, note that mov instruction has no re-relativization problems, so the fix is easy to implement), but I hope this error can be fixed on the official side. This small error may cause serious consequents as the execution results will be inconsistency between the binary level trace and the original exe execution.

**Comment by derekbruening** [comment]
DR hooks certain syscall wrappers and kernel entry functions in ntdll.dll.  The hook uses a jump instruction.  You are reading the jump instruction, when you want the displaced code.   You can see how the drmemtrace tracing framework handles this by encoding the instruction instead of copying from the result of instr_get_app_pc(): https://github.com/DynamoRIO/dynamorio/blob/master/clients/drcachesim/tracer/instru_offline.cpp#L510.  This is not documented in the interface so it is not surprising it is confusing.

The hook trampoline resumption sequence is hidden by dr_fragment_app_pc() translating from the interception buffer via get_app_pc_from_intercept_pc_if_necessary(), but there is no mechanism to hide the hook if a client looks at the app pc.  The instr_get_app_pc() needs to point at the original PC values, including the hook jump, in order for translation to work properly, so that is not changeable.  Probably the action item is to document the suggested solution that drmemtrace uses, in the instr_get_app_pc() docs maybe.

Please also see how drmemtrace handles emulation, enabling handling of rep string, scatter/gather, and other sequences which are typically rewritten by tracing tools.  If all you need is an instruction trace with encodings, you may find that drmemtrace with `-instr_only_trace` works out of the box and you may not need a custom client.

**Comment by yangyixiaof** [comment]
Yes, thanks for your quick response and help! OK, I look forward to see the documentation updation.

**Comment by yangyixiaof** [comment]
I will close this issue if the documentation is updated.

**Comment by yangyixiaof** [comment]
I learned something new from the code link you provide, very thanks. I think these important materials should be put somewhere in the documentation.

**Comment by yangyixiaof** [comment]
> DR hooks certain syscall wrappers and kernel entry functions in ntdll.dll. The hook uses a jump instruction. You are reading the jump instruction, when you want the displaced code. You can see how the drmemtrace tracing framework handles this by encoding the instruction instead of copying from the result of instr_get_app_pc(): https://github.com/DynamoRIO/dynamorio/blob/master/clients/drcachesim/tracer/instru_offline.cpp#L510. This is not documented in the interface so it is not surprising it is confusing.
> 
> The hook trampoline resumption sequence is hidden by dr_fragment_app_pc() translating from the interception buffer via get_app_pc_from_intercept_pc_if_necessary(), but there is no mechanism to hide the hook if a client looks at the app pc. The instr_get_app_pc() needs to point at the original PC values, including the hook jump, in order for translation to work properly, so that is not changeable. Probably the action item is to document the suggested solution that drmemtrace uses, in the instr_get_app_pc() docs maybe.
> 
> Please also see how drmemtrace handles emulation, enabling handling of rep string, scatter/gather, and other sequences which are typically rewritten by tracing tools. If all you need is an instruction trace with encodings, you may find that drmemtrace with `-instr_only_trace` works out of the box and you may not need a custom client.


Very thanks, the materials in your comment are really helpful and I learn something new. I think these materials should be put somewhere in the documentation. Thanks for the link which contains the handling of emulation code and its original instr. However, I found that for emulated AVX scatter/gather or rep instr(rep stos, etc), the emulated code contains short jump, and as these happen in instrumentation phase 2, DynamoRIO cannot split a whole block in phase 2 into multi blocks which contain long jump, am I right? For a DynamoRIO block, only the last instr is a jmp or syscall. So there is a restriction when instrumenting the emulated code, that is the clean call (save fpstate) should be avoided as a clean call take many instructions which may exceed the short jump distance on Windows. These detailed materials are not in the documentation and I found the restriction by doing many experiments by myself. Thus, I choose to emulate the rep and scatter/gather in the clean call by myself. Thanks again for the drmemtrace advice, I am implementating a simple selective dynamic symbolic execution on Windows, the symbolic engine to use is Triton, thus, despite the original binary trace, I also need the concrete values, so, I need a new client, but still thanks for the advice, the implementation from official side is really thorough. 

Besides, I am curious about how DynamoRIO handles the last jmp instr jumps to a block which is intrumented and stored at somewhere else? I mean if original jmp instr is jmp A, A is an original app_pc, and the block at A is instrumented and stored at B, when does DynamoRIO replace the jmp A to jmp B? Does the replace happen in phase 4 or phase 5? where can I find the related materials to read? I searched the developpper materials, there is no details about that.

**Comment by derekbruening** [comment]
> I found that for emulated AVX scatter/gather or rep instr(rep stos, etc), the emulated code contains short jump, and because these happen in instrumentation phase 2, DynamoRIO cannot split a block (containing short jmp) into multi blocks which contain long jump after phase 2 , am I right? For a DynamoRIO block, only the last instr is a long jmp or syscall. So there is a restriction when instrumenting the emulated code, that is the clean call (save fpstate) should be avoided because a clean call take many instructions which may exceed the short jump distance on Windows. These detailed materials are not in the documentation only mentioned a little in the dr_insert_clean_call API doc and I ensure the restriction by doing many experiments by myself. Thus, I choose to emulate the rep and scatter/gather in the inserted clean call by myself. 

You mean you inserted clean calls in the middle of the emulation regions: between "jmp iter" and "iter:" for rep string, e.g.  No other tool had hit any problem there including very heavyweight tools like Dr. Memory and drmemtrace: probably because there is only a move-immediate and a jump in between there and most tools wouldn't need a clean call to handle those.

Please file a new issue on the short jump problem you hit with these emulation expansions.  While we do not recommend using clean calls for every single instruction, these expansion libraries should have a way to work with such instrumentation, whether that's permanently switching to long jumps or having an option to use long jumps.  Better to get these libraries to work, than for you to duplicate the emulation.

**Comment by derekbruening** [comment]
> Besides, I am curious about how DynamoRIO handles the last jmp instr jumps to a block which is intrumented and stored at somewhere else? I mean if the original jmp instr is jmp A, A is an original app_pc, and the block at A is instrumented and stored at B, when does DynamoRIO replace the jmp A to jmp B? Does the replace happen in phase 4 or phase 5? where can I find the related materials to read? I searched the developpper guides, there is no details about that.

It sounds like you are asking about the base operation of dynamic code discovery and copying into blocks in the code cache.  One reference is "Efficient, Transparent, and Comprehensive Runtime Code Manipulation." linked at https://dynamorio.org/page_publications.html

**Comment by yangyixiaof** [comment]
> > I found that for emulated AVX scatter/gather or rep instr(rep stos, etc), the emulated code contains short jump, and because these happen in instrumentation phase 2, DynamoRIO cannot split a block (containing short jmp) into multi blocks which contain long jump after phase 2 , am I right? For a DynamoRIO block, only the last instr is a long jmp or syscall. So there is a restriction when instrumenting the emulated code, that is the clean call (save fpstate) should be avoided because a clean call take many instructions which may exceed the short jump distance on Windows. These detailed materials are not in the documentation only mentioned a little in the dr_insert_clean_call API doc and I ensure the restriction by doing many experiments by myself. Thus, I choose to emulate the rep and scatter/gather in the inserted clean call by myself.
> 
> You mean you inserted clean calls in the middle of the emulation regions: between "jmp iter" and "iter:" for rep string, e.g. No other tool had hit any problem there including very heavyweight tools like Dr. Memory and drmemtrace: probably because there is only a move-immediate and a jump in between there and most tools wouldn't need a clean call to handle those.
> 
> Please file a new issue on the short jump problem you hit with these emulation expansions. While we do not recommend using clean calls for every single instruction, these expansion libraries should have a way to work with such instrumentation, whether that's permanently switching to long jumps or having an option to use long jumps. Better to get these libraries to work, than for you to duplicate the emulation.

Yes, if more than three  clean calls are inserted between "jmp iter" and "iter:", it will exceed the short jump max distance on Windows. I have used clean call for every instruction, as I need to record whether the instr is executed or not for conditional instruction and the timestamp this instr is executed (plus the concrete values of reg or mem when the instr is executed). Additionally, I have done a simple taint analysis and branch judgement in the execution phase, so I almost use the clean call for every instruction. That works fine for currrent tested exes. Thanks for your excellent work! I pay tribute to your work.  I want to open source my code (a demo) after my current academic work is done. By the way, any of my work is far from having as much influence and practicality as yours. I can't imagine that DynamoRIO is just a doctoral work done by you (a single PhD at that time). Besides, I read your PhD thesis, That is really an excellent paper and work. Regards to you again.

**Comment by derekbruening** [comment]
As mentioned, if you could file a new separate issue on the short jump problem you hit with these emulation expansions, to avoid confusing this issue which is on instr_get_app_pc().

**Comment by yangyixiaof** [comment]
> As mentioned, if you could file a new separate issue on the short jump problem you hit with these emulation expansions, to avoid confusing this issue which is on instr_get_app_pc().

OK, I will open a new issue describing the short jump problem.
